src/main/java/com/mraof/minestuck/item/weapon/RightClickBlockEffect.java

```diff
@@ -1,11 +1,78 @@
 package com.mraof.minestuck.item.weapon;
 
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
 import net.minecraft.item.ItemUseContext;
-import net.minecraft.util.ActionResultType;
+import net.minecraft.util.*;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.BlockRayTraceResult;
+import net.minecraft.world.World;
+
+import java.util.function.Supplier;
 
 public interface RightClickBlockEffect
 {
 	ActionResultType onClick(ItemUseContext context);
 	
+	static RightClickBlockEffect placeFluid(Supplier<Block> fluidBlock, Supplier<Item> otherItem)
+	{
+		return (context) -> {
+			World worldIn = context.getWorld();
+			PlayerEntity player = context.getPlayer();
+			ItemStack itemStack = context.getItem();
+			Direction facing = context.getFace();
+			BlockPos pos = context.getPos().offset(facing);
+			
+			if(!worldIn.isRemote && player != null)
+			{
+				BlockState state = worldIn.getBlockState(pos);
+				
+				if(state.getBlock() == Blocks.AIR || state.getBlock() == fluidBlock.get())
+				{
+					worldIn.setBlockState(pos, fluidBlock.get().getDefaultState());
+					ItemStack newItem = new ItemStack(otherItem.get(), itemStack.getCount());
+					newItem.setTag(itemStack.getTag()); //It is important that the item it is switching to has the same durability
+					player.setHeldItem(context.getHand(), newItem);
+					worldIn.playSound(null, pos, SoundEvents.ITEM_BUCKET_EMPTY, SoundCategory.BLOCKS, 1F, 2F);
+					player.getCooldownTracker().setCooldown(otherItem.get(), 5);
+					return ActionResultType.SUCCESS;
+				}
+			}
+			return ActionResultType.PASS;
+		};
+	}
 	
+	static RightClickBlockEffect scoopBlock(Supplier<Block> validBlock)
+	{
+		return (context) -> {
+			World worldIn = context.getWorld();
+			BlockPos pos = context.getPos();
+			PlayerEntity player = context.getPlayer();
+			Direction facing = context.getFace();
+			boolean inside = context.isInside();
+			
+			if(!worldIn.isRemote && player != null)
+			{
+				BlockState state = worldIn.getBlockState(pos);
+				
+				if(state.getBlock() == validBlock.get().getBlock())
+				{
+					ItemStack block = state.getPickBlock(new BlockRayTraceResult(context.getHitVec(), facing, pos, inside), worldIn, pos, player);
+					player.inventory.addItemStackToInventory(block);
+					context.getItem().damageItem(1, player, (playerEntity) -> playerEntity.sendBreakAnimation(Hand.MAIN_HAND));
+					
+					worldIn.playSound(null, pos, SoundEvents.ENTITY_ITEM_PICKUP, SoundCategory.NEUTRAL, 1F, 1F);
+					worldIn.setBlockState(pos, Blocks.AIR.getDefaultState());
+					
+					return ActionResultType.SUCCESS;
```

Comment:
It's a good idea to return success/pass independent of `worldIn.isRemote`. As such, I'd suggest switching the order of `!worldIn.isRemote` and `state.getBlock() == validBlock.get().getBlock()`, and then return success if the block is valid independently of if the world is remote or not.

Reply:
Now it's doing a bit much client-side, though it's not too obvious. There can currently be some desyncing if this effect is performed client-side, but something gets in the way so that it doesn't perform server-side.

Testing it myself (changed the code to not do anything here for `!worldIn.isRemote`),
- `addItemStackToInventory()` should be done server-side only (can cause ghost items)
- `setBlockState()` might be good to have server-side only (desync is short-lived, but still present). Also is server-side only in `HoeItem.onItemUse()`, so might be good to follow their example
- `damageItem()` not sure if this can cause any sensible desync, but based on `HoeItem.onItemUse()`, should probably be server-side only
- `playSound()` should probably be both-sides, however it's technically server-side only because you're passing in a null player (pass in the player as the argument and it'll play the sound directly client-side for the player using the weapon)

Reply:
Server-side specificity made